synapse/third_party_rules/access_rules.py

```diff
@@ -0,0 +1,375 @@
+    to define custom rules for specific events and actions.
+    Implements the custom behaviour for the "im.vector.room.access_rules" state event.
+
+    Takes a config in the format:
+
+    third_party_event_rules:
+        module: third_party_rules.RoomAccessRules
+        config:
+            # List of domains (server names) that can't be invited to rooms if the
+            # "restricted" rule is set. Defaults to an empty list.
+            domains_forbidden_when_restricted: []
+
+            # Identity server to use when checking the HS an email address belongs to
+            # using the /info endpoint. Required.
+            id_server: "vector.im"
+
+    Don't forget to consider if you can invite users from your own domain.
+    """
+
+    def __init__(self, config, http_client):
+        self.http_client = http_client
+
+        self.id_server = config["id_server"]
+
+        self.domains_forbidden_when_restricted = config.get(
+            "domains_forbidden_when_restricted", [],
+        )
+
+    @staticmethod
+    def parse_config(config):
+        if "id_server" in config:
+            return config
+        else:
+            raise ConfigError("No IS for event rules TchapEventRules")
+
+    def on_create_room(self, requester, config, is_requester_admin):
+        """Implements synapse.events.ThirdPartyEventRules.on_create_room
+
+        Checks if a im.vector.room.access_rules event is being set during room creation.
+        If yes, make sure the event is correct. Otherwise, append an event with the
+        default rule to the initial state.
+        """
+        is_direct = config.get("is_direct")
+        rules_in_initial_state = False
+
+        # If there's a rules event in the initial state, check if it complies with the
+        # spec for im.vector.room.access_rules and fix it if not.
+        for event in config.get("initial_state", []):
+            if event["type"] == ACCESS_RULES_TYPE:
+                rules_in_initial_state = True
+
+                rule = event["content"].get("rule")
+
+                # Make sure the event has a valid content.
+                if rule is None:
+                    event["content"] = {
+                        "rule": self._on_create_room_default_rule(is_direct)
+                    }
+
+                # Make sure the rule name is valid.
+                if not self._is_rule_name_valid(rule):
+                    event["content"]["rule"] = self._on_create_room_default_rule(
+                        is_direct,
+                    )
+
+                # Make sure the rule is "direct" if the room is a direct chat.
+                if is_direct and rule != ACCESS_RULE_DIRECT:
+                    event["content"]["rule"] = ACCESS_RULE_DIRECT
+
+                # Make sure the rule is not "direct" if the room isn't a direct chat.
+                if rule == ACCESS_RULE_DIRECT and not is_direct:
+                    event["content"]["rule"] = ACCESS_RULE_RESTRICTED
```

Comment:
I'd deny in all these cases, as the event is invalid, rather than trying to fix it up for the clients (and possibly doing the wrong thing).